### PR TITLE
Lab jaren

### PR DESCRIPTION
**Pull request by jjron**
This was a tough, interesting assignment, took about 8 hours with around an extra 2 for the bonus.
It gave a lot of good practice for finding the docs that might be needed. I initially used a forEach for the bonus to iterate through filePaths, which would return the text data in-order or out-of-order with every other run and pass every other time mocha runs. I installed the async module to try their asynchronous eachOf but I still get the same result. Looking forward to seeing the answer tomorrow.

### CODE REVIEW

File: lab-jaren/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+### Using done in Mocha callbacks
```
Comment (JudyVue) [inline lab-jaren/README.md:1]: Great README description.

File: lab-jaren/lib/read-files.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+const fs = require('fs');
+
+module.exports = function(filePaths, callback) {
+  var result = [];
+  fs.readFile(filePaths[0], (err, data) => {
```
Comment (JudyVue) [inline lab-jaren/lib/read-files.js:7]: Great job using arrow functions.

File: lab-jaren/test/read-files-test.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+const expect = require('chai').expect;
+const readFiles = require('../lib/read-files.js');
+
+describe('testing module readFiles', function(){
+  describe('with a file that does not exist', function(){
+    it('should pass an error into the callback', function(done){
+      readFiles([`${__dirname}/../assets/does-not-exist.txt`], function(err, data){
+        //console.log(err);
+        expect(Boolean(err)).to.equal(true);
```
Comment (JudyVue) [inline lab-jaren/test/read-files-test.js:11]: Nice job on making the error a Boolean! Duncan usually teaches us to make the syntax```expect(!!err).to.equal(true)``` when testing if an error occurs, but yours also works really well.

File: lab-jaren/test/read-files-test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        'Open up in there! Oh, no! There isn\'t any other way out. I can\'t hold them off forever! Now what? This is some rescue. When you came in here, didn\'t you have a plan for getting out? He\'s the brains, sweetheart. Well, I didn\'t... What the hell are you doing? Somebody has to save our skins. Into the garbage chute, wise guy. Get in there you big furry oaf! I don\'t care what you smell! Get in there and don\'t worry about it. Wonderful girl! Either I\'m going to kill her or I\'m beginning to like her. Get in there!\n',
+        'What? Yahoo! Look out! You\'re all clear, kid. Now let\'s blow this thing and go home! Stand by to fire at Rebel base. Standing by. Great shot, kid. That was one in a million. Remember, the Force will be with you...always.\n' ];
+      readFiles(inputFiles, function(err, data){
```
Comment (JudyVue) [inline lab-jaren/test/read-files-test.js:27]: Between lines 27-28,  you want to have some kind of error handling in case your ```readFiles``` function fails. So before your ```expect``` statements, put: 

```   if (err) return done(err); ```

File: lab-jaren/test/read-files-test.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    });
+  });
+
+  describe('with valid inputs', function(){
+    var inputFiles = [`${__dirname}/../assets/star.txt`,
+      `${__dirname}/../assets/wars.txt`,
+      `${__dirname}/../assets/ipsum.txt`,
+    ];
+    it('should pass an array of text data that maps from the file paths array', function(done){
+      var result = [ 'Well, that\'s the trick, isn\'t it? And it\'s going to cost you something extra. Ten thousand in advance. Ten thousand? We could almost buy our own ship for that! But who\'s going to fly it, kid! You? You bet I could. I\'m not such a bad pilot myself! We don\'t have to sit here and listen... We haven\'t that much with us. But we could pay you two thousand now, plus fifteen when we reach Alderaan. Seventeen, huh! Okay. You guys got yourself a ship. We\'ll leave as soon as you\'re ready. Docking bay Ninety-four. Ninety-four. Looks like somebody\'s beginning to take an interest in your handiwork. All right, we\'ll check it out.\n',
```
Comment (JudyVue) [inline lab-jaren/test/read-files-test.js:24]: Thomas and I had a hard time figuring out how to avoid copy/pasting ALL the text to test the proper outputs. In the end, Duncan helped him (and me too :) with a DRYer version of the test module. 

It's still NOT the dryest it can be (it's more or less repeating the nested fs.readFile functions in a before block), but it's still better-looking than copy/pasting text. You can look at Thomas's test module in his PR here:

[https://github.com/codefellows-seattle-javascript-401d14/lab-03-readfiles/pull/4/files](https://github.com/codefellows-seattle-javascript-401d14/lab-03-readfiles/pull/4/files)

I hope this helps!

File: lab-jaren/test/read-many-files-test.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+'use strict';
```
Comment (JudyVue) [inline lab-jaren/test/read-many-files-test.js:1]: Great job going above and beyond attempting the bonus AND writing tests for it. I don't have an answer to solving the async issue yet, but definitely go to Duncan if you're interested in solving it.

File: lab-jaren/index.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+const main = module.exports = function(argv){
+  if (!argv || argv.length < 3) return 'Usage error: must provide at least one file path';
+  return readManyFiles(argv.slice(2, argv.length), function(err, data){
+    if (err) return console.error(err);
+    console.log(data);
+  });
+};
+
+console.log(main(process.argv));
```
Comment (JudyVue) [inline lab-jaren/index.js:15]: You're not the only one who did this, and this doesn't break your code, but here's a copy/pasted comment I gave to Kenneth because had the same thing. 

========================

Just a small thing I wasn't aware of until I ran the code on my machine. Let's get rid of the ```console.log``` and just invoke the ```main``` function so that it says ```main(process.argv)```. 

When I ran your code in my terminal, I was first getting ```undefined``` and *then* the array of text output. If we just get rid of ```console.log``` and invoke ```main(process.argv)``` alone, we still get text output printed out because we already log the data on Line 12. 

The reason  for that is the console isn't able to log the output from ```main(process.argv)``` because the ```main``` function is async. 

In other words, ```console.log``` happens instantly while ```main(process.argv)``` is still going through the process of completing your ```readfiles``` function that happens on Line 8. 

```Console.log``` isn't going to wait until ```main(process.argv)``` is done, it's going to just print out contents right away, but since ```readFiles``` on Line 8 still hasn't completed, ```console.log``` spits out ```undefined.```

If you're confused, come find me and I'll try to re-explain again. :)

File: lab-jaren/lib/read-many-files.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+const fs = require('fs');
+const async = require('async');
```
Comment (JudyVue) [inline lab-jaren/lib/read-many-files.js:4]: You did some serious google-fu to find this cool package. Awesome job at being so resourceful.

File: lab-jaren/lib/read-many-files.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+module.exports = function(filePaths, callback) {
+  var result = [];
+  async.eachOf(filePaths, function(path, index){
```
Comment (JudyVue) [inline lab-jaren/lib/read-many-files.js:8]: Good job figuring out how to use this new module's syntax. I can totally see you studied the shit out of those docs. 

Again, as to controlling the async, I'm not too sure myself, but it's a good thing to bring up with Duncan when you have time. 
 
I think you did a great job figuring this out either way, in spite of the async issue.

File: lab-jaren/lib/read-many-files.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    fs.readFile(path, (err, data) => {
+      if (err) return callback(err);
+      result.push(data.toString());
+      if (index === filePaths.length-1)
```
Comment (JudyVue) [inline lab-jaren/lib/read-many-files.js:12]: I'm curious, what is this line
```if (index === filePaths.length-1)```
 meant to accomplish?

Comment (jjron) [reply lab-jaren/lib/read-many-files.js:12]: I was trying to call the callback function with result only on the last iteration of readFile, and async's eachOf function has an index param to keep track of the index as it loops. I was thinking if filePaths.length-1 matches the index of the last path in the array, then I should invoke the callback.

File: lab-jaren/lib/read-files.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    if (err) return callback(err);
+    result.push(data.toString());
+    fs.readFile(filePaths[1], (err, data) => {
+      if (err) return callback(err);
+      result.push(data.toString());
+      fs.readFile(filePaths[2], (err, data) => {
+        if (err) return callback(err);
+        result.push(data.toString());
+        callback(null, result);
+      });
```
Comment (jjron) [inline lab-jaren/lib/read-files.js:17]: In the async.eachOf, this is why I wanted to ensure the callback only invoked on the last iteration.
